pkg/rate/api_limiter.go

```diff
@@ -0,0 +1,794 @@
+		goto skipRateLimiter
+	}
+
+	if l.params.MinWaitDuration > 0 && limitWaitDuration < l.params.MinWaitDuration {
+		limitWaitDuration = l.params.MinWaitDuration
+	}
+
+	if l.params.MaxWaitDuration > 0 && limitWaitDuration > l.params.MaxWaitDuration {
+		if l.params.Log {
+			scopedLog.Warning("Not processing API request. Wait duration exceeds maximum")
+		}
+		return nil, fmt.Errorf("request would have to wait %v to be served (maximum wait duration: %v)",
+			limitWaitDuration, l.params.MaxWaitDuration)
+	}
+
+	if limitWaitDuration != 0 {
+		select {
+		case <-time.After(limitWaitDuration):
+		case <-ctx.Done():
+			if l.params.Log {
+				scopedLog.Warning("Not processing API request due to cancelled context while waiting")
+			}
+			return nil, fmt.Errorf("request cancelled while waiting for rate limiting slot: %w", ctx.Err())
+		}
+	}
+
+	if parallelRequests > 0 {
+		waitCtx := ctx
+		if l.params.MaxWaitDuration > 0 {
+			maxWaitDurationParallel := l.params.MaxWaitDuration - limitWaitDuration
+			scopedLog = scopedLog.WithField(logMaxWaitDurationParallel, maxWaitDurationParallel)
+			ctx2, cancel := context.WithTimeout(ctx, maxWaitDurationParallel)
+			defer cancel()
+			waitCtx = ctx2
+		}
+		waitSemaphoreWeight = int64(waitSemaphoreResolution / parallelRequests)
+		err := l.parallelWaitSemaphore.Acquire(waitCtx, waitSemaphoreWeight)
+		if err != nil {
+			if l.params.Log {
+				scopedLog.Warning("Not processing API request. Wait duration for maximum parallel requests exceeds maximum")
+			}
+			return nil, fmt.Errorf("timed out while waiting to be served with %d parallel requests: %w", parallelRequests, err)
+		}
+	}
+
+	waitDuration = time.Since(scheduledAt)
+
+skipRateLimiter:
+
+	l.mutex.Lock()
+	l.currentRequestsInFlight++
+	l.mutex.Unlock()
+
+	log.WithFields(logrus.Fields{
+		logAPICallName:       l.name,
+		logUUID:              uuid,
+		logWaitDurationTotal: waitDuration,
+	}).Debug("API call is ready to be served")
+
+	if l.params.Log {
+		scopedLog.Info("Processing API request with rate limiter")
+	} else {
+		scopedLog.Debug("Processing API request with rate limiter")
+	}
+
+	return &limitedRequest{
+		limiter:             l,
+		startTime:           time.Now(),
+		scheduleTime:        scheduledAt,
+		waitDuration:        waitDuration,
+		waitSemaphoreWeight: waitSemaphoreWeight,
+		uuid:                uuid,
+	}, nil
+}
+
+func parseRate(r string) (rate.Limit, error) {
+	tokens := strings.SplitN(r, "/", 2)
+	if len(tokens) != 2 {
+		return 0, fmt.Errorf("not in the form number/interval")
+	}
+
+	f, err := strconv.ParseFloat(tokens[0], 64)
+	if err != nil {
+		return 0, fmt.Errorf("unable to parse float %q: %w", tokens[0], err)
+	}
+
+	// Reject rates such as 1/1 or 10/10 as it will default to nanoseconds
+	// which is likely unexpected to the user. Require an explicit suffix.
+	if _, err := strconv.ParseInt(string(tokens[1]), 10, 64); err == nil {
+		return 0, fmt.Errorf("interval %q must contain duration suffix", tokens[1])
+	}
```

Comment:
:facepalm: sorry, I re-read the code and realised that the `ParseDuration()` case below will handle this already, returning an error like `time: missing unit in duration "1"`.  I just didn't see it in the tests so I assumed that the user could pass a value like this.

Anyway, minor nit.